### PR TITLE
Factory Design Pattern and Facade Design Pattern

### PR DESCRIPTION
**Pull request by nazrinzuwair**
Hi @ash-teach 

I've created PR #77, which includes updates for both the Factory Design Pattern and the Facade Design Pattern:

1. Created folders named "Factory-Design-Pattern" and "Facade-Pattern" inside the PATTERN-LIBRARY directory. Each folder contains a README.md file and JavaScript files (factory-method.js and facade-pattern.js).
2. Added a README.md file in the "Factory-Design-Pattern" folder with a summary of implementing the Factory Design Pattern in JavaScript using the factory design method. Included a code snippet example related to "Creating Foods" and listed additional programming languages where the Factory Design Pattern can be implemented.
3. Included a README.md file in the "Facade-Pattern" folder with explanations and code examples for implementing the Facade Design Pattern in JavaScript.
4. Added a reference section in the README.md files of both folders with links to websites browsed during research.
5. Linked the "Factory-Design-Pattern" and "Facade-Pattern" folders into the README.md main file for easy navigation.

### CODE REVIEW

**Comment by nazrinzuwair** [comment]
@ash-teach  Could you please review this.

**Comment by kattsky** [comment]
Hi @nazrinzuwair I've reviewed your code, and overall, it's very well-structured and maintains good object-oriented principles. I also like to commend your detailed documentation of your implementation that made it easy to grasp the goal and functionality of the facade pattern.

Building upon the foundation of a well-documented code, one aspect to consider exploring to enhance flexibility is create instances of the subsystem classes outside of the `SoftwareFacade` class,.

As an alternative to creating instances of `SoftwareDeveloper`, `Designer`, and `DatabaseEngineer` within the `SoftwareFacade` constructor, these can be passed as parameters. The benefit of creating instances of the subsystem classes externally rather than internal implementation helps `SoftwareFacade` object (being the mediator between the client code and the subsystem classes) to access them in the software development process which aligns with facade pattern's goal to provide simplified interface to a complex system.

:raised_hands: Fantastic work, @nazrinzuwair

**Comment by nazrinzuwair** [comment]
Thanks for your feedback @kattsky :)

File: patterns/Facade-Pattern/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+class DatabaseEngineer {
+    // Define a method within the DatabaseEngineer class to represent connecting
+    connect() {
+      console.log('Database engineer performs the connection between the application and the database.');
+    }
+}
+```
+
+We have three classes here: SoftwareDeveloper, Designer, and DatabaseEngineer, each of which represents a subsystem responsible for a specific task in the software development process. The SoftwareDeveloper class's code() method takes handle of coding activities, the Designer class's design() method takes responsibility of designing tasks, and the DatabaseEngineer class's connect() method takes charge of database connection tasks. Each subsystem's behavior is specified by these classes.
+
```
Comment (ash-teach) [inline patterns/Facade-Pattern/README.md:53]: Any time you reference code in a sentence or paragraph you can use in-line code markdown (i.e. single backticks) to denote that you are referring to code. For example, the paragraph above should be styled as follows:

We have three classes here: `SoftwareDeveloper`, `Designer`, and `DatabaseEngineer`, each of which represents a subsystem responsible for a specific task in the software development process. The `SoftwareDeveloper` class's `code()`method takes handle of coding activities, the `Designer` class's `design()` method takes responsibility of designing tasks, and the `DatabaseEngineer` class's `connect()` method takes charge of database connection tasks. Each subsystem's behavior is specified by these classes.

Comment (ash-teach) [reply patterns/Facade-Pattern/README.md:53]: Oh! I see you've done that below - you must have just missed this paragraph.

File: patterns/Factory-Design-Patten/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+![javascript-factory-method](https://github.com/nazrinzuwair/pattern-library/assets/144160072/f5ab27dc-bd49-4c0b-9fed-92bce6f7df4d)
+
+The following entities are involved in this pattern when implementing using JavaScript:
+
+   - Creator: This corresponds to the 'factory' object in the provided visual representation, responsible for producing new products. It implements the 'factoryMethod', which yields freshly generated products.
+
+   - Concrete Product: These are the products being created, represented in the example code by the 'Employees' objects. Each product adheres to the same interface, comprising identical attributes and operations.
+
+**Abstract Product: This role is not commonly used in JavaScript. Traditionally, it would declare a product interface for various programming languages or scenarios. However, JavaScript does not specify interfaces explicitly. Instead, actual objects typically implement the relevant properties and methods directly, eliminating the need for an abstract product interface.**
+
```
Comment (ash-teach) [inline patterns/Factory-Design-Patten/README.md:18]: Did you intended to style the above paragraph all in **bold** and not part of the list with Creator and Concrete Product?  This just seems like an odd styling choice, and I'm not sure what you intend to communicate by it.

File: patterns/Factory-Design-Patten/README.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+(Poorte, 2023)
+
+## Example demonstrating the Factory Method pattern in JavaScript:
+
+### Factory Function for Creating Foods
+
+The FoodFactory function serves as a factory for creating different types of foods, demonstrating the Factory Design Pattern. It contains a method createFood that takes a type as input and returns a corresponding food object. This approach abstracts the process of object creation, allowing the client code to create food objects without knowing the specific type being instantiated.
+
```
Comment (ash-teach) [inline patterns/Factory-Design-Patten/README.md:26]: As with the comment above - please make sure all inline reference to code in sentences or paragraphs are appropriately styled with single backticks.

File: patterns/Factory-Design-Patten/README.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        let food;
+        
+        // Create different foods based on the type provided
+        if (type === "pizza") {
+            food = new Pizza();
+        } else if (type === "burger") {
+            food = new Burger();
+        } else if (type === "salad") {
+            food = new Salad();
+        }
```
Comment (ash-teach) [inline patterns/Factory-Design-Patten/README.md:47]: Maybe a [switch statement](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch) is a semantic improvement here?

File: patterns/Factory-Design-Patten/README.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+}
+```
+### Pizza Class 
+
+The Pizza class represents a type of food created by the FoodFactory factory function. It encapsulates the properties and behavior specific to pizza objects. By being created through the FoodFactory, the Pizza class adheres to the Factory Design Pattern, where the factory method (createFood) is responsible for creating instances of concrete classes (Pizza) without exposing the instantiation logic.
+
+```javascript
+// Pizza Class
+function Pizza() {
+    this.famousAudience = "youngers!";
```
Comment (ash-teach) [inline patterns/Factory-Design-Patten/README.md]: "young**sters**!" :)

File: patterns/Factory-Design-Patten/README.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+// Burger Class
+function Burger() {
+    this.famousAudience = "Kids!";
+}
+```
+
+### Salad Class 
+
+The Salad class represents yet another type of food created by the FoodFactory factory function. It encapsulates the properties and behavior specific to salad objects. As with the Pizza and Burger classes, the instantiation of Salad objects through the FoodFactory adheres to the Factory Design Pattern, promoting encapsulation and decoupling by separating the client code from the object creation logic.
+
```
Comment (ash-teach) [inline patterns/Factory-Design-Patten/README.md]: While I appreciate your inclusion of these details in each food paragraph, I think you could have one paragraph to list the benefits attributable to all three classes, and remove the paragraphs for each food. That will cut down on unnecessary repetition.

**Review by ash-teach — CHANGES_REQUESTED** [review]

**Comment by ash-teach** [comment]
Great work, @nazrinzuwair! Very well documented and structured. I have very few comments, other than primarily some styling related notes. Though I do want to highlight @kattsky's comment as valid (quoted below), and worth considering.

> As an alternative to creating instances of `SoftwareDeveloper`, `Designer`, and `DatabaseEngineer` within the `SoftwareFacade` constructor, these can be passed as parameters. The benefit of creating instances of the subsystem classes externally rather than internal implementation helps `SoftwareFacade` object (being the mediator between the client code and the subsystem classes) to access them in the software development process which aligns with facade pattern's goal to provide simplified interface to a complex system.

As a further note, I would normally suggest that you do not need to reproduce all code as part of your documentation (the user can easily visit the actual code afterall). However, in this case, you've done such a great job of discussing the code, I would prefer to leave it as-is. Fantastic work!

**Comment by nazrinzuwair** [comment]
Hi @kattsky and @ash-teach 
Thanks for your valuable feedback. I have made the changes and it is ready for your review. 
Thank You!

**Comment by kattsky** [comment]
Thanks for catching those details, @nazrinzuwair. Got nothing else to add from my end.

**Pull request merged**